File: src/main/java/pl/com/tegess/config/security/AdminService.java
package pl.com.tegess.config.security;

import org.mongodb.morphia.query.Query;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import pl.com.tegess.config.Config;
import pl.com.tegess.domain.admin.Admin;
import pl.com.tegess.domain.admin.AdminRepository;

import java.util.Optional;

@Configuration
@Import({Config.class})
public class AdminService implements UserDetailsService {

    @Autowired
    AdminRepository adminRepository;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        Query<Admin> query = adminRepository.createQuery().field("username").equal(username);
        return Optional.ofNullable(adminRepository.findOne(query)).orElseThrow(
                () -> new UsernameNotFoundException("This user doesn't exists!"));
    }
}

File: src/main/scala/com/tegess/persistance/service/admin/AdminService.scala
package com.tegess.persistance.service.admin

import com.tegess.domain.admin.Admin
import com.tegess.persistance.MongoConfig
import com.tegess.persistance.service.user.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.{Configuration, Import}
import org.springframework.security.core.userdetails.{UserDetails, UserDetailsService}

import scala.util.{Failure, Success, Try}

@Configuration
@Import(Array(classOf[MongoConfig]))
class AdminService @Autowired() (userService: UserService) extends UserDetailsService {

  override def loadUserByUsername(username: String): UserDetails = {

    val result = for {
      applicationId <- Try(Admin.getApplicationIdFromUsername(username))
      realUsername <- Try(Admin.getRealUsernameFromUsername(username))
      user <- Try(userService.findOne(applicationId, realUsername).get)
    } yield Admin(user)

    result match {
      case Success(admin) => admin
      case Failure(e) => throw e
    }
  }
}
